source/adapters/level_zero/v2/command_list_cache.hpp

```diff
@@ -0,0 +1,63 @@
+//===--------- command_list_cache.hpp - Level Zero Adapter ---------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <stack>
+
+#include <ur/ur.hpp>
+#include <ur_api.h>
+#include <ze_api.h>
+
+#include "common.hpp"
+
+namespace v2 {
+struct immediate_command_list_descriptor_t {
+  ze_device_handle_t Device;
+  ZeStruct<ze_command_queue_desc_t> QueueDesc;
+  bool operator==(const immediate_command_list_descriptor_t &rhs) const;
+};
+
+struct regular_command_list_descriptor_t {
+  ze_device_handle_t Device;
+  bool IsInOrder;
+  uint32_t Ordinal;
+  bool operator==(const regular_command_list_descriptor_t &rhs) const;
+};
+
+using command_list_descriptor_t =
+    std::variant<immediate_command_list_descriptor_t,
+                 regular_command_list_descriptor_t>;
+
+struct command_list_descriptor_hash_t {
+  inline size_t operator()(const command_list_descriptor_t &desc) const;
+};
+
+struct command_list_cache_t {
+  command_list_cache_t(
+      std::function<void(ze_command_list_handle_t)> Destructor =
+          [](ze_command_list_handle_t cmdList) {
+            ZE_CALL_NOCHECK(zeCommandListDestroy, (cmdList));
+          });
+  ~command_list_cache_t();
+
+  ze_command_list_handle_t
+  getCommandList(const command_list_descriptor_t &desc);
+  void addCommandList(const command_list_descriptor_t &desc,
+                      ze_command_list_handle_t cmdList);
+
+private:
+  std::unordered_map<command_list_descriptor_t,
+                     std::stack<ze_command_list_handle_t>,
```

Comment:
maybe wrap `ze_command_list_handle_t` in a struct/class with a destructor? will eliminate the function pointer you are storing.

Reply:
Done. The callback was only introduced so that I could easily test the cache without actually creating proper command lists but this is now outdated.